File: c++/driver.cpp
#include "matrix.h"
#include "losses.h"
#include "layers.h"
#include "metrics.h"

using namespace matrix;

int main() {
    // Lets setup our data
    std::vector<std::vector<float>> in { 
	                       {1.0, 2.0, 3.0},
			       {-4.0, -5.0, -6.0},  
			       {7.0, 8.0, 9.0} };
    
    std::vector<std::vector<float>> true_in {{0, 1},
					   {1, 0},
					   {0, 1}};
    Matrix<float> X(3, 3);
    X.set_matrix(in);
    Matrix<float> y_true(3, 2);
    y_true.set_matrix(true_in);

    Dense layer1(3, 3);
    ReLU layer2;
    Dense layer3(3, 2);
    SoftmaxCrossEntropy layer4;
    print(X);
    Matrix<float> out1 = layer1.forward(X);
    print(out1);
    Matrix<float> out2 = layer2.forward(out1);
    print(out2);
    Matrix<float> out3 = layer3.forward(out2);
    print(out3);
    std::cout << "Start of softmax \n";
    Matrix<float> out4 = layer4.forward(out3, y_true);
    print(out4);
    std::cout << layer4.get_loss() << std::endl;
    std::cout << metric::accuracy(y_true, out4) << "\n";

    std::cout << "Start of backprop relu and dense" << std::endl;

    Matrix<float> back1 = layer2.backward(out2);
    print(back1);
    Matrix<float> back2 = layer1.backward(back1);
    print(back2);
    print(layer1.get_dbiases());
    print(layer1.get_dweights());

    std::cout << "START OF TEST" << std::endl;

    std::vector<std::vector<float>> test_in {{0.7, 0.1, 0.2},
					   {0.1, 0.5, 0.4},
					   {0.02, 0.9, 0.08}};
    Matrix<float> softmaxOut(3, 3);
    softmaxOut.set_matrix(test_in);
    Matrix<float> test5 = layer4.backward(softmaxOut, y_true);
    print(test5);
    std::cout << "Start of backprop full" << std::endl;


    return 1;
}

File: python/driver.py
from Layers import *
from Losses import *
from Metrics import *
from Optimizers import *
from data.data import *

# Load our data
x_train = np.array(x_train_raw_1000)
y_train = np.array(y_train_raw_1000)
x_test = np.array(x_test_raw_1000)
y_test = np.array(y_test_raw_1000)

# Define out network
layer1 = Dense(2, 16)
layer2 = ReLU()
layer3 = Dense(16, 2)
layer4 = SoftmaxCrossEntropy()
sgd = SGD(learning_rate = 1, decay=0.001)

# Perform training loop
for epoch in range(2001):
    # Forward Pass
    out = layer1.forward(x_train)
    out2 = layer2.forward(out)
    out3 = layer3.forward(out2)
    out4 = layer4.forward(out3, y_train)

    # Calculate loss and metric
    loss = layer4.loss
    acc = Accuracy().calculate(y_train, out4)

    # Backward Pass
    back4 = layer4.backward(out4, y_train)
    back3 = layer3.backward(back4)
    back3 = layer2.backward(back3)
    back1 = layer1.backward(back3)

    # Update parameters
    sgd.pre_update()
    sgd.update(layer3)
    sgd.update(layer1)
    sgd.post_update()

    # Test the network every 100 iterations
    if not epoch % 100:
        out = layer1.forward(x_test)
        out2 = layer2.forward(out)
        out3 = layer3.forward(out2)
        out4 = layer4.forward(out3, y_test)
        lossTest = layer4.loss
        accTest = Accuracy().calculate(y_test, out4)
        
        print(f'epoch: {epoch}, ' +
        f'acc: {acc:.3f}, ' +
        f'loss: {loss:.3f}, ' +
        f'acc_test: {accTest:.3f}, ' +
        f'loss_test: {lossTest:.3f}, ' +
        f'lr: {sgd.current_learning_rate:.3f}')
